WorkflowTest.shouldCompleteWorkflowInstanceWithVariables flaky

**Summary**
- How often does the test fail? at least once so far
- Does it block your work? no
- Do we suspect that it is a real failure? unknown

**Failures**

<details><summary>Example assertion failure</summary>
<pre>
io.zeebe.client.api.command.ClientStatusException: deadline exceeded after 1.999764931s. [remote_addr=0.0.0.0/127.0.0.1:27704]
	at io.zeebe.client.impl.ZeebeClientFutureImpl.transformExecutionException(ZeebeClientFutureImpl.java:93)
	at io.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:50)
	at io.zeebe.test.WorkflowTest.shouldCompleteWorkflowInstanceWithVariables(WorkflowTest.java:56)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeLazy(JUnitCoreWrapper.java:119)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:87)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:157)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
Caused by: java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1.999764931s. [remote_addr=0.0.0.0/127.0.0.1:27704]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1999)
	at io.zeebe.client.impl.ZeebeClientFutureImpl.join(ZeebeClientFutureImpl.java:48)
	... 43 more
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1.999764931s. [remote_addr=0.0.0.0/127.0.0.1:27704]
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:460)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426)
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$1CloseInContext.runInContext(ClientCallImpl.java:416)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
</pre>
</details>

**Hypotheses**


**Logs**

<details><summary>Logs</summary>
<pre>
[10:47:49.273 [] [main] INFO  io.zeebe.test - Test finished: shouldCompleteWorkflowInstance(io.zeebe.test.WorkflowTest)
10:47:49.273 [] [main] INFO  io.zeebe.test - Test started: shouldCompleteWorkflowInstanceWithVariables(io.zeebe.test.WorkflowTest)
10:47:49.274 [] [main] INFO  io.zeebe.test.util.SocketUtil - Choosing next port 27704 for test fork 2 with range PortRange{host='localhost', basePort=27700, maxOffset=100, currentOffset=5}
10:47:49.275 [] [main] INFO  io.zeebe.test.util.SocketUtil - Choosing next port 27705 for test fork 2 with range PortRange{host='localhost', basePort=27700, maxOffset=100, currentOffset=6}
10:47:49.275 [] [main] INFO  io.zeebe.test.util.SocketUtil - Choosing next port 27706 for test fork 2 with range PortRange{host='localhost', basePort=27700, maxOffset=100, currentOffset=7}
10:47:49.275 [] [main] INFO  io.zeebe.test.util.SocketUtil - Choosing next port 27707 for test fork 2 with range PortRange{host='localhost', basePort=27700, maxOffset=100, currentOffset=8}
10:47:49.276 [] [main] DEBUG io.zeebe.broker.system - Initializing system with base path /tmp/6e5cee6d-b755-4580-a310-70708550e326
10:47:49.279 [] [main] INFO  io.zeebe.broker.system - Version: 0.25.0-SNAPSHOT
10:47:49.281 [] [main] INFO  io.zeebe.broker.system - Starting broker 0 with configuration {
  "network" : {
    "host" : "0.0.0.0",
    "portOffset" : 0,
    "maxMessageSize" : "4MB",
    "advertisedHost" : "0.0.0.0",
    "commandApi" : {
      "host" : "0.0.0.0",
      "port" : 27705,
      "advertisedHost" : "0.0.0.0",
      "advertisedPort" : 27705,
      "address" : "0.0.0.0:27705",
      "advertisedAddress" : "0.0.0.0:27705"
    },
    "internalApi" : {
      "host" : "0.0.0.0",
      "port" : 27706,
      "advertisedHost" : "0.0.0.0",
      "advertisedPort" : 27706,
      "address" : "0.0.0.0:27706",
      "advertisedAddress" : "0.0.0.0:27706"
    },
    "monitoringApi" : {
      "host" : "0.0.0.0",
      "port" : 27707,
      "advertisedHost" : "0.0.0.0",
      "advertisedPort" : 27707,
      "address" : "0.0.0.0:27707",
      "advertisedAddress" : "0.0.0.0:27707"
    },
    "maxMessageSizeInBytes" : 4194304
  },
  "cluster" : {
    "initialContactPoints" : [ ],
    "partitionIds" : [ 1 ],
    "nodeId" : 0,
    "partitionsCount" : 1,
    "replicationFactor" : 1,
    "clusterSize" : 1,
    "clusterName" : "zeebe-cluster",
    "membership" : {
      "broadcastUpdates" : false,
      "broadcastDisputes" : true,
      "notifySuspect" : false,
      "gossipInterval" : "PT0.25S",
      "gossipFanout" : 2,
      "probeInterval" : "PT1S",
      "probeTimeout" : "PT2S",
      "suspectProbes" : 3,
      "failureTimeout" : "PT10S",
      "syncInterval" : "PT10S"
    }
  },
  "threads" : {
    "cpuThreadCount" : 2,
    "ioThreadCount" : 2
  },
  "data" : {
    "directories" : [ "/tmp/6e5cee6d-b755-4580-a310-70708550e326/data" ],
    "logSegmentSize" : "512MB",
    "snapshotPeriod" : "PT15M",
    "logIndexDensity" : 100,
    "diskUsageReplicationWatermark" : 0.9,
    "diskUsageCommandWatermark" : 0.8,
    "diskUsageMonitoringInterval" : "PT1S",
    "logSegmentSizeInBytes" : 536870912,
    "atomixStorageLevel" : "DISK",
    "freeDiskSpaceCommandWatermark" : 0,
    "freeDiskSpaceReplicationWatermark" : 0
  },
  "exporters" : {
    "test-recorder" : {
      "jarPath" : null,
      "className" : "io.zeebe.test.util.record.RecordingExporter",
      "args" : null,
      "external" : false
    }
  },
  "gateway" : {
    "network" : {
      "host" : "0.0.0.0",
      "port" : 27704,
      "minKeepAliveInterval" : "PT30S"
    },
    "cluster" : {
      "contactPoint" : "0.0.0.0:27706",
      "requestTimeout" : "PT15S",
      "clusterName" : "zeebe-cluster",
      "memberId" : "gateway",
      "host" : "0.0.0.0",
      "port" : 26502,
      "membership" : {
        "broadcastUpdates" : false,
        "broadcastDisputes" : true,
        "notifySuspect" : false,
        "gossipInterval" : "PT0.25S",
        "gossipFanout" : 2,
        "probeInterval" : "PT1S",
        "probeTimeout" : "PT2S",
        "suspectProbes" : 3,
        "failureTimeout" : "PT10S",
        "syncInterval" : "PT10S"
      }
    },
    "threads" : {
      "managementThreads" : 1
    },
    "monitoring" : {
      "enabled" : false,
      "host" : "0.0.0.0",
      "port" : 9600
    },
    "security" : {
      "enabled" : false,
      "certificateChainPath" : null,
      "privateKeyPath" : null
    },
    "longPolling" : {
      "enabled" : true
    },
    "initialized" : true,
    "enable" : true
  },
  "backpressure" : {
    "enabled" : true,
    "algorithm" : "VEGAS",
    "aimd" : {
      "requestTimeout" : "PT1S",
      "initialLimit" : 100,
      "minLimit" : 1,
      "maxLimit" : 1000,
      "backoffRatio" : 0.9
    },
    "fixedLimit" : {
      "limit" : 20
    },
    "vegas" : {
      "alpha" : 3,
      "beta" : 6,
      "initialLimit" : 20
    },
    "gradient" : {
      "minLimit" : 10,
      "initialLimit" : 20,
      "rttTolerance" : 2.0
    },
    "gradient2" : {
      "minLimit" : 10,
      "initialLimit" : 20,
      "rttTolerance" : 2.0,
      "longWindow" : 600
    }
  },
  "stepTimeout" : "PT5M",
  "executionMetricsExporterEnabled" : false
}
10:47:49.282 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [1/13]: actor scheduler
10:47:49.283 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [1/13]: actor scheduler started in 1 ms
10:47:49.283 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [2/13]: membership and replication protocol
10:47:49.289 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [2/13]: membership and replication protocol started in 6 ms
10:47:49.289 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [3/13]: command api transport
10:47:49.296 [] [main] DEBUG io.zeebe.broker.system - Bound command API to 0.0.0.0:27705 
10:47:49.296 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [3/13]: command api transport started in 7 ms
10:47:49.296 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [4/13]: command api handler
10:47:49.297 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [4/13]: command api handler started in 0 ms
10:47:49.297 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [5/13]: subscription api
10:47:49.298 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [5/13]: subscription api started in 1 ms
10:47:49.298 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [6/13]: embedded gateway
10:47:49.298 [] [main] INFO  io.zeebe.gateway - Version: 0.25.0-SNAPSHOT
10:47:49.298 [] [main] INFO  io.zeebe.gateway - Starting gateway with configuration {
  "network" : {
    "host" : "0.0.0.0",
    "port" : 27704,
    "minKeepAliveInterval" : "PT30S"
  },
  "cluster" : {
    "contactPoint" : "0.0.0.0:27706",
    "requestTimeout" : "PT15S",
    "clusterName" : "zeebe-cluster",
    "memberId" : "gateway",
    "host" : "0.0.0.0",
    "port" : 26502,
    "membership" : {
      "broadcastUpdates" : false,
      "broadcastDisputes" : true,
      "notifySuspect" : false,
      "gossipInterval" : "PT0.25S",
      "gossipFanout" : 2,
      "probeInterval" : "PT1S",
      "probeTimeout" : "PT2S",
      "suspectProbes" : 3,
      "failureTimeout" : "PT10S",
      "syncInterval" : "PT10S"
    }
  },
  "threads" : {
    "managementThreads" : 1
  },
  "monitoring" : {
    "enabled" : false,
    "host" : "0.0.0.0",
    "port" : 9600
  },
  "security" : {
    "enabled" : false,
    "certificateChainPath" : null,
    "privateKeyPath" : null
  },
  "longPolling" : {
    "enabled" : true
  },
  "initialized" : true,
  "enable" : true
}
10:47:49.303 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [6/13]: embedded gateway started in 5 ms
10:47:49.303 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [7/13]: cluster services
10:47:49.356 [] [raft-server-0-raft-partition-partition-1] WARN  io.atomix.utils.event.ListenerRegistry - Listener io.atomix.raft.roles.FollowerRole$$Lambda$324/0x00000008403f4c40@75aea90c not registered
10:47:49.386 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [7/13]: cluster services started in 82 ms
10:47:49.386 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [8/13]: topology manager
10:47:49.386 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [8/13]: topology manager started in 0 ms
10:47:49.387 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [9/13]: monitoring services
10:47:49.387 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [9/13]: monitoring services started in 0 ms
10:47:49.387 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [10/13]: disk space monitor
10:47:49.388 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [10/13]: disk space monitor started in 0 ms
10:47:49.388 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [11/13]: leader management request handler
10:47:49.388 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [11/13]: leader management request handler started in 0 ms
10:47:49.388 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [12/13]: zeebe partitions
10:47:49.388 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 partitions [1/1]: partition 1
10:47:49.389 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Removing follower partition service for partition PartitionId{id=1, group=raft-partition}
10:47:49.389 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Partition role transitioning from null to LEADER
10:47:49.390 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Installing leader partition service for partition PartitionId{id=1, group=raft-partition}
10:47:49.561 [GatewayTopologyManager] [Broker-0-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 0, partitions {} and terms {}.
10:47:55.920 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Opened database from '/tmp/6e5cee6d-b755-4580-a310-70708550e326/data/raft-partition/partitions/1/runtime'.
10:47:56.001 [Broker-0-LogStream-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams - Configured log appender back pressure at partition 1 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, alphaLimit=0.7, betaLimit=0.95}. Window limiting is disabled
10:47:56.009 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams - Recovering state of partition 1 from snapshot
10:47:56.009 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.zeebe.el.impl.feel.MessagePackValueMapper@3bf2befe)), function-provider: io.zeebe.el.impl.feel.FeelFunctionProvider@beeb65c, configuration: Configuration(false)]
10:47:56.019 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  io.zeebe.logstreams - Recovered state of partition 1 from snapshot at position -1
10:47:56.020 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.zeebe.el.impl.feel.MessagePackValueMapper@4b964779)), function-provider: io.zeebe.el.impl.feel.FeelFunctionProvider@65bd15bb, configuration: Configuration(false)]
10:47:56.021 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.zeebe.el.impl.feel.MessagePackValueMapper@1b1f0306)), function-provider: io.zeebe.el.impl.feel.FeelFunctionProvider@4ebeb661, configuration: Configuration(false)]
10:47:56.035 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Recovering exporter from snapshot
10:47:56.036 [Broker-0-HealthCheckService] [Broker-0-zb-actors-0] DEBUG io.zeebe.broker.system - All partitions are installed. Broker is ready!
10:47:56.038 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 partitions [1/1]: partition 1 started in 6649 ms
10:47:56.038 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 partitions succeeded. Started 1 steps in 6650 ms.
10:47:56.038 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [12/13]: zeebe partitions started in 6650 ms
10:47:56.038 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 [13/13]: register diskspace usage listeners
10:47:56.038 [] [main] DEBUG io.zeebe.broker.system - Bootstrap Broker-0 [13/13]: register diskspace usage listeners started in 0 ms
10:47:56.038 [] [main] INFO  io.zeebe.broker.system - Bootstrap Broker-0 succeeded. Started 13 steps in 6756 ms.
10:47:56.048 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Recovered exporter 'Broker-0-Exporter-1' from snapshot at lastExportedPosition -1
10:47:56.048 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Configure exporter with id 'test-recorder'
10:47:56.048 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Set event filter for exporters: ExporterEventFilter{acceptRecordTypes={NULL_VAL=true, COMMAND_REJECTION=true, SBE_UNKNOWN=true, EVENT=true, COMMAND=true}, acceptValueTypes={NULL_VAL=true, VARIABLE_DOCUMENT=true, ERROR=true, VARIABLE=true, WORKFLOW_INSTANCE_CREATION=true, WORKFLOW_INSTANCE_SUBSCRIPTION=true, TIMER=true, MESSAGE_SUBSCRIPTION=true, JOB=true, WORKFLOW_INSTANCE_RESULT=true, JOB_BATCH=true, MESSAGE_START_EVENT_SUBSCRIPTION=true, MESSAGE=true, INCIDENT=true, DEPLOYMENT=true, WORKFLOW_INSTANCE=true, SBE_UNKNOWN=true}}
10:47:56.049 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Open exporter with id 'test-recorder'
10:47:56.067 [GatewayTopologyManager] [Broker-0-zb-actors-0] DEBUG io.zeebe.gateway - Received metadata change from Broker 0, partitions {1=LEADER} and terms {1=1}.
10:47:56.139 [] [main] INFO  io.zeebe.test - 
====
Broker startup time: 6863
====

10:47:59.944 [] [main] INFO  io.zeebe.test.records - Test failed, following records were exported:
10:47:59.960 [] [main] INFO  io.zeebe.test.records - {"valueType":"DEPLOYMENT","key":-1,"position":1,"timestamp":1594637276158,"recordType":"COMMAND","intent":"CREATE","partitionId":1,"rejectionType":"NULL_VAL","rejectionReason":"","value":{"resources":[{"resource":"PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPGJwbW46ZGVmaW5pdGlvbnMgeG1sbnM6YnBtbj0iaHR0cDovL3d3dy5vbWcub3JnL3NwZWMvQlBNTi8yMDEwMDUyNC9NT0RFTCIgeG1sbnM6YnBtbmRpPSJodHRwOi8vd3d3Lm9tZy5vcmcvc3BlYy9CUE1OLzIwMTAwNTI0L0RJIiB4bWxuczpkaT0iaHR0cDovL3d3dy5vbWcub3JnL3NwZWMvREQvMjAxMDA1MjQvREkiIHhtbG5zOmRjPSJodHRwOi8vd3d3Lm9tZy5vcmcvc3BlYy9ERC8yMDEwMDUyNC9EQyIgeG1sbnM6eHNpPSJodHRwOi8vd3d3LnczLm9yZy8yMDAxL1hNTFNjaGVtYS1pbnN0YW5jZSIgeG1sbnM6emVlYmU9Imh0dHA6Ly9jYW11bmRhLm9yZy9zY2hlbWEvemVlYmUvMS4wIiBpZD0iRGVmaW5pdGlvbnNfMSIgdGFyZ2V0TmFtZXNwYWNlPSJodHRwOi8vYnBtbi5pby9zY2hlbWEvYnBtbiIgZXhwb3J0ZXI9IkNhbXVuZGEgTW9kZWxlciIgZXhwb3J0ZXJWZXJzaW9uPSIxLjguMiI+CiAgPGJwbW46cHJvY2VzcyBpZD0icHJvY2VzcyIgaXNFeGVjdXRhYmxlPSJ0cnVlIj4KICAgIDxicG1uOnN0YXJ0RXZlbnQgaWQ9InN0YXJ0IiBuYW1lPSJTdGFydCI+CiAgICAgIDxicG1uOm91dGdvaW5nPl...
10:47:59.961 [] [main] INFO  io.zeebe.test.records - {"valueType":"DEPLOYMENT","key":2251799813685250,"position":2,"timestamp":1594637276872,"recordType":"EVENT","intent":"CREATED","partitionId":1,"rejectionType":"NULL_VAL","rejectionReason":"","value":{"resources":[{"resource":"PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLTgiPz4KPGJwbW46ZGVmaW5pdGlvbnMgeG1sbnM6YnBtbj0iaHR0cDovL3d3dy5vbWcub3JnL3NwZWMvQlBNTi8yMDEwMDUyNC9NT0RFTCIgeG1sbnM6YnBtbmRpPSJodHRwOi8vd3d3Lm9tZy5vcmcvc3BlYy9CUE1OLzIwMTAwNTI0L0RJIiB4bWxuczpkaT0iaHR0cDovL3d3dy5vbWcub3JnL3NwZWMvREQvMjAxMDA1MjQvREkiIHhtbG5zOmRjPSJodHRwOi8vd3d3Lm9tZy5vcmcvc3BlYy9ERC8yMDEwMDUyNC9EQyIgeG1sbnM6eHNpPSJodHRwOi8vd3d3LnczLm9yZy8yMDAxL1hNTFNjaGVtYS1pbnN0YW5jZSIgeG1sbnM6emVlYmU9Imh0dHA6Ly9jYW11bmRhLm9yZy9zY2hlbWEvemVlYmUvMS4wIiBpZD0iRGVmaW5pdGlvbnNfMSIgdGFyZ2V0TmFtZXNwYWNlPSJodHRwOi8vYnBtbi5pby9zY2hlbWEvYnBtbiIgZXhwb3J0ZXI9IkNhbXVuZGEgTW9kZWxlciIgZXhwb3J0ZXJWZXJzaW9uPSIxLjguMiI+CiAgPGJwbW46cHJvY2VzcyBpZD0icHJvY2VzcyIgaXNFeGVjdXRhYmxlPSJ0cnVlIj4KICAgIDxicG1uOnN0YXJ0RXZlbnQgaWQ9InN0YXJ0IiBuYW1lPSJTdGFydCI+CiAgICAgIDxicG1uO...
10:47:59.964 [] [main] INFO  io.zeebe.test - Test execution time: 3825
10:47:59.964 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [1/13]: register diskspace usage listeners
10:47:59.964 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [1/13]: register diskspace usage listeners closed in 0 ms
10:47:59.964 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [2/13]: zeebe partitions
10:47:59.964 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 partitions [1/1]: partition 1
10:47:59.964 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-1] DEBUG io.zeebe.broker.exporter - Closed exporter director 'Broker-0-Exporter-1'.
10:47:59.965 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.exporter.stream.ExporterDirector@78f32ca8 successfully
10:47:59.965 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$608/0x0000000840568440@3c49d7c5 successfully
10:47:59.965 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.impl.AsyncSnapshotDirector@36bfb508 successfully
10:47:59.965 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$389/0x0000000840431040@2b01f5cf successfully
10:47:59.966 [Broker-0-StreamProcessor-1] [Broker-0-zb-actors-0] DEBUG io.zeebe.logstreams - Closed stream processor controller Broker-0-StreamProcessor-1.
10:47:59.966 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.engine.processor.StreamProcessor@4a761a6b successfully
10:47:59.966 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$401/0x0000000840434c40@38f31b1f successfully
10:47:59.966 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.logstreams.LogDeletionService@73affefc successfully
10:48:00.003 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.logstreams.snapshot - Closed database from '/tmp/6e5cee6d-b755-4580-a310-70708550e326/data/raft-partition/partitions/1/runtime'.
10:48:00.003 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$399/0x0000000840435440@57a4d2cb successfully
10:48:00.003 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$391/0x0000000840431840@2dfb2f16 successfully
10:48:00.003 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$390/0x0000000840431440@d70e255 successfully
10:48:00.003 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$389/0x0000000840431040@124b519 successfully
10:48:00.003 [Broker-0-LogStream-1] [Broker-0-zb-actors-1] INFO  io.zeebe.logstreams - Close appender for log stream logstream-raft-partition-partition-1
10:48:00.004 [logstream-raft-partition-partition-1-write-buffer] [Broker-0-zb-actors-1] DEBUG io.zeebe.dispatcher - Dispatcher closed
10:48:00.004 [Broker-0-LogStream-1] [Broker-0-zb-actors-1] INFO  io.zeebe.logstreams - On closing logstream logstream-raft-partition-partition-1 close 3 readers
10:48:00.004 [Broker-0-LogStream-1] [Broker-0-zb-actors-1] INFO  io.zeebe.logstreams - Close log storage with name logstream-raft-partition-partition-1
10:48:00.004 [Broker-0-ZeebePartition-1] [Broker-0-zb-actors-1] DEBUG io.zeebe.broker.system - Closed io.zeebe.broker.system.partitions.ZeebePartition$$Lambda$387/0x0000000840430840@70e71d7c successfully
10:48:00.005 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 partitions [1/1]: partition 1 closed in 41 ms
10:48:00.005 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 partitions succeeded. Closed 1 steps in 41 ms.
10:48:00.005 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [2/13]: zeebe partitions closed in 41 ms
10:48:00.005 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [3/13]: leader management request handler
10:48:00.005 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [3/13]: leader management request handler closed in 0 ms
10:48:00.005 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [4/13]: disk space monitor
10:48:00.005 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [4/13]: disk space monitor closed in 0 ms
10:48:00.005 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [5/13]: monitoring services
10:48:00.006 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [5/13]: monitoring services closed in 0 ms
10:48:00.006 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [6/13]: topology manager
10:48:00.006 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [6/13]: topology manager closed in 0 ms
10:48:00.006 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [7/13]: cluster services
10:48:00.006 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [7/13]: cluster services closed in 0 ms
10:48:00.006 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [8/13]: embedded gateway
10:48:00.007 [] [main] DEBUG io.zeebe.gateway - Closing gateway broker client ...
10:48:00.008 [] [main] DEBUG io.zeebe.gateway - topology manager closed
10:48:00.008 [] [main] DEBUG io.zeebe.gateway - Gateway broker client closed.
10:48:00.008 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [8/13]: embedded gateway closed in 2 ms
10:48:00.008 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [9/13]: subscription api
10:48:00.009 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [9/13]: subscription api closed in 0 ms
10:48:00.009 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [10/13]: command api handler
10:48:00.009 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [10/13]: command api handler closed in 0 ms
10:48:00.009 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [11/13]: command api transport
10:48:00.017 [io.zeebe.gateway.impl.broker.BrokerRequestManager] [Broker-0-zb-actors-1] ERROR io.zeebe.gateway - Expected to handle gRPC request, but an unexpected error occurred
io.zeebe.gateway.impl.broker.RequestRetriesExhaustedException: Expected to execute the command on one of the partitions, but all failed; there are no more partitions available to retry. Please try again. If the error persists contact your zeebe operator
	at io.zeebe.gateway.impl.broker.RequestRetryHandler.sendRequestWithRetry(RequestRetryHandler.java:128) ~[zeebe-gateway-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.gateway.impl.broker.RequestRetryHandler.lambda$sendRequestWithRetry$1(RequestRetryHandler.java:115) ~[zeebe-gateway-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
	at io.zeebe.gateway.impl.broker.BrokerRequestManager.lambda$sendRequestInternal$2(BrokerRequestManager.java:132) ~[zeebe-gateway-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.invoke(ActorJob.java:76) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorJob.execute(ActorJob.java:39) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorTask.execute(ActorTask.java:118) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:107) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.doWork(ActorThread.java:91) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	at io.zeebe.util.sched.ActorThread.run(ActorThread.java:204) [zeebe-util-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
	Suppressed: java.net.ConnectException
		at io.atomix.cluster.messaging.impl.AbstractClientConnection.close(AbstractClientConnection.java:123) ~[atomix-cluster-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
		at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$getOrCreateClientConnection$30(NettyMessagingService.java:578) ~[atomix-cluster-0.25.0-SNAPSHOT.jar:0.25.0-SNAPSHOT]
		at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:577) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:551) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:490) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:615) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.util.concurrent.DefaultPromise.setSuccess0(DefaultPromise.java:604) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.util.concurrent.DefaultPromise.trySuccess(DefaultPromise.java:104) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.channel.DefaultChannelPromise.trySuccess(DefaultChannelPromise.java:84) ~[netty-transport-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.channel.AbstractChannel$CloseFuture.setClosed(AbstractChannel.java:1158) ~[netty-transport-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.channel.AbstractChannel$AbstractUnsafe.doClose0(AbstractChannel.java:760) ~[netty-transport-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:736) ~[netty-transport-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.channel.AbstractChannel$AbstractUnsafe.close(AbstractChannel.java:607) ~[netty-transport-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.channel.epoll.AbstractEpollChannel$AbstractEpollUnsafe.shutdownInput(AbstractEpollChannel.java:484) ~[netty-transport-native-epoll-4.1.51.Final-linux-x86_64.jar:4.1.51.Final]
		at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:815) ~[netty-transport-native-epoll-4.1.51.Final-linux-x86_64.jar:4.1.51.Final]
		at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:475) ~[netty-transport-native-epoll-4.1.51.Final-linux-x86_64.jar:4.1.51.Final]
		at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378) ~[netty-transport-native-epoll-4.1.51.Final-linux-x86_64.jar:4.1.51.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.51.Final.jar:4.1.51.Final]
		at java.lang.Thread.run(Thread.java:834) ~[?:?]
10:48:02.170 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [11/13]: command api transport closed in 2161 ms
10:48:02.171 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [12/13]: membership and replication protocol
10:48:08.064 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [12/13]: membership and replication protocol closed in 5892 ms
10:48:08.064 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 [13/13]: actor scheduler
10:48:08.066 [] [main] DEBUG io.zeebe.broker.system - Closing Broker-0 [13/13]: actor scheduler closed in 2 ms
10:48:08.066 [] [main] INFO  io.zeebe.broker.system - Closing Broker-0 succeeded. Closed 13 steps in 8102 ms.
10:48:08.066 [] [main] INFO  io.zeebe.broker.system - Broker shut down.
10:48:08.104 [] [main] INFO  io.zeebe.test - Broker closing time: 8140
</pre>
</details>

--- a/test/pom.xml
+++ b/test/pom.xml
@@ -1,4 +1,6 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
 
   <modelVersion>4.0.0</modelVersion>
   <name>Zeebe Test</name>
@@ -56,6 +58,11 @@
     <dependency>
       <groupId>io.zeebe</groupId>
       <artifactId>zeebe-protocol</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>io.zeebe</groupId>
+      <artifactId>zeebe-logstreams</artifactId>
     </dependency>
 
     <dependency>

--- a/test/src/main/java/io/zeebe/test/EmbeddedBrokerRule.java
+++ b/test/src/main/java/io/zeebe/test/EmbeddedBrokerRule.java
@@ -8,8 +8,10 @@
 package io.zeebe.test;
 
 import static io.zeebe.test.util.TestUtil.waitUntil;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import io.zeebe.broker.Broker;
+import io.zeebe.broker.PartitionListener;
 import io.zeebe.broker.SpringBrokerBridge;
 import io.zeebe.broker.system.EmbeddedGatewayService;
 import io.zeebe.broker.system.configuration.BrokerCfg;
@@ -18,6 +20,7 @@ import io.zeebe.broker.system.configuration.NetworkCfg;
 import io.zeebe.gateway.impl.broker.BrokerClient;
 import io.zeebe.gateway.impl.broker.cluster.BrokerClusterState;
 import io.zeebe.gateway.impl.broker.cluster.BrokerTopologyManager;
+import io.zeebe.logstreams.log.LogStream;
 import io.zeebe.test.util.TestConfigurationFactory;
 import io.zeebe.test.util.record.RecordingExporter;
 import io.zeebe.test.util.record.RecordingExporterTestWatcher;
@@ -26,11 +29,16 @@ import io.zeebe.util.FileUtil;
 import io.zeebe.util.ZbLogger;
 import io.zeebe.util.allocation.DirectBufferAllocator;
 import io.zeebe.util.sched.clock.ControlledActorClock;
+import io.zeebe.util.sched.future.ActorFuture;
+import io.zeebe.util.sched.future.CompletableActorFuture;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.InetSocketAddress;
+import java.time.Duration;
 import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import org.assertj.core.util.Files;
@@ -49,13 +57,14 @@ public class EmbeddedBrokerRule extends ExternalResource {
   private static final String STATE_DIRECTORY = "state";
   protected final RecordingExporterTestWatcher recordingExporterTestWatcher =
       new RecordingExporterTestWatcher();
-  protected final BrokerCfg brokerCfg;
-  protected final ControlledActorClock controlledActorClock = new ControlledActorClock();
   protected final Supplier<InputStream> configSupplier;
   protected final Consumer<BrokerCfg>[] configurators;
+  protected BrokerCfg brokerCfg;
   protected Broker broker;
+  protected final ControlledActorClock controlledActorClock = new ControlledActorClock();
+  protected final SpringBrokerBridge springBrokerBridge = new SpringBrokerBridge();
   protected long startTime;
-  private final int timeout;
+  private final Duration timeout;
   private final File newTemporaryFolder;
   private List<String> dataDirectories;
 
@@ -80,6 +89,14 @@ public class EmbeddedBrokerRule extends ExternalResource {
   public EmbeddedBrokerRule(
       final Supplier<InputStream> configSupplier,
       final int timeout,
+      final Consumer<BrokerCfg>... configurators) {
+    this(configSupplier, Duration.ofSeconds(timeout), configurators);
+  }
+
+  @SafeVarargs
+  public EmbeddedBrokerRule(
+      final Supplier<InputStream> configSupplier,
+      final Duration timeout,
       final Consumer<BrokerCfg>... configurators) {
     this.configSupplier = configSupplier;
     this.configurators = configurators;
@@ -128,7 +145,7 @@ public class EmbeddedBrokerRule extends ExternalResource {
   }
 
   @Override
-  protected void before() {
+  public void before() {
     startTime = System.currentTimeMillis();
     startBroker();
     LOG.info("\n====\nBroker startup time: {}\n====\n", (System.currentTimeMillis() - startTime));
@@ -136,7 +153,7 @@ public class EmbeddedBrokerRule extends ExternalResource {
   }
 
   @Override
-  protected void after() {
+  public void after() {
     try {
       LOG.info("Test execution time: " + (System.currentTimeMillis() - startTime));
       startTime = System.currentTimeMillis();
@@ -155,6 +172,8 @@ public class EmbeddedBrokerRule extends ExternalResource {
       } catch (final IOException e) {
         LOG.error("Unexpected error on deleting data.", e);
       }
+
+      controlledActorClock.reset();
     }
   }
 
@@ -167,7 +186,7 @@ public class EmbeddedBrokerRule extends ExternalResource {
   }
 
   public Broker getBroker() {
-    return this.broker;
+    return broker;
   }
 
   public ControlledActorClock getClock() {
@@ -193,8 +212,24 @@ public class EmbeddedBrokerRule extends ExternalResource {
             brokerCfg,
             newTemporaryFolder.getAbsolutePath(),
             controlledActorClock,
-            new SpringBrokerBridge());
+            springBrokerBridge);
+
+    final CountDownLatch latch = new CountDownLatch(brokerCfg.getCluster().getPartitionsCount());
+    broker.addPartitionListener(new LeaderPartitionListener(latch));
+
     broker.start().join();
+
+    try {
+      final boolean hasLeaderPartition = latch.await(timeout.toMillis(), TimeUnit.MILLISECONDS);
+
+      assertThat(hasLeaderPartition)
+          .describedAs("Expected the broker to have a leader of the partition within %s", timeout)
+          .isTrue();
+
+    } catch (final InterruptedException e) {
+      LOG.info("Timeout. Broker was not started within {}", timeout, e);
+      Thread.currentThread().interrupt();
+    }
 
     final EmbeddedGatewayService embeddedGatewayService = broker.getEmbeddedGatewayService();
     if (embeddedGatewayService != null) {
@@ -239,6 +274,27 @@ public class EmbeddedBrokerRule extends ExternalResource {
           deleteSnapshots(stateDirectory);
         }
       }
+    }
+  }
+
+  private static class LeaderPartitionListener implements PartitionListener {
+
+    private final CountDownLatch latch;
+
+    LeaderPartitionListener(final CountDownLatch latch) {
+      this.latch = latch;
+    }
+
+    @Override
+    public ActorFuture<Void> onBecomingFollower(final int partitionId, final long term) {
+      return CompletableActorFuture.completed(null);
+    }
+
+    @Override
+    public ActorFuture<Void> onBecomingLeader(
+        final int partitionId, final long term, final LogStream logStream) {
+      latch.countDown();
+      return CompletableActorFuture.completed(null);
     }
   }
 }

